### PR TITLE
feat: Skip the records in appointments having jsr validation for user profile with appointment file load (RDCC-1565)

### PR DESCRIPTION
**Pull request by abhijitdiwan**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDCC-1565

### Change description ###
feat: Skip the records in appointments having jsr validation for user profile with appointment file load (RDCC-1565)

**Does this PR introduce a breaking change?** (check one with "x")
No

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/JudicialUserProfileProcessor.java [hunk 1]
```diff
@@ -34,10 +36,20 @@ public void process(Exchange exchange) {
         List<JudicialUserProfile> filteredJudicialUserProfiles = validate(judicialUserProfileJsrValidatorInitializer,
                 judicialUserProfiles);
 
+        //Set List of Elinks which are skipped due to JSR violations in CSV ROW
+        filteredUserProfileKeys = judicialUserProfiles.stream()
+                .filter(profileList -> (filteredJudicialUserProfiles.stream().noneMatch(filteredList ->
+                        filteredList.getElinksId().equalsIgnoreCase(profileList.getElinksId()))))
+                .map(map -> map.getElinksId()).collect(toList());
+
```
[the remark this answers is not in the document]
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/JudicialUserProfileProcessor.java]: done

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/JudicialUserProfileProcessor.java [hunk 2]
```diff
@@ -34,10 +36,20 @@ public void process(Exchange exchange) {
         List<JudicialUserProfile> filteredJudicialUserProfiles = validate(judicialUserProfileJsrValidatorInitializer,
                 judicialUserProfiles);
 
+        //Set List of Elinks which are skipped due to JSR violations in CSV ROW
+        filteredUserProfileKeys = judicialUserProfiles.stream()
+                .filter(profileList -> (filteredJudicialUserProfiles.stream().noneMatch(filteredList ->
+                        filteredList.getElinksId().equalsIgnoreCase(profileList.getElinksId()))))
```
[the remark this answers is not in the document]
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/JudicialUserProfileProcessor.java]: done

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/JudicialOfficeAppointmentProcessor.java [hunk 3]
```diff
@@ -35,8 +41,32 @@ public void process(Exchange exchange) {
...
+     * Filters valid user profile elinks id records and ignore invalid user profile elinks record with JSR errors.
+     *
+     * @param filteredJudicialAppointments List
+     * @return List JudicialOfficeAppointment
+     */
+    private List<JudicialOfficeAppointment> filterValidAppointmentsForUserProfile(List<JudicialOfficeAppointment> filteredJudicialAppointments) {
+        if (nonNull(judicialUserProfileProcessor.getFilteredUserProfileKeys())) {
+            filteredJudicialAppointments = filteredJudicialAppointments.stream().filter(appointments ->
+                    judicialUserProfileProcessor.getFilteredUserProfileKeys().stream()
+                            .noneMatch(skippedIdParent -> appointments.getElinksId().equalsIgnoreCase(skippedIdParent)))
```
[the remark this answers is not in the document]
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/JudicialOfficeAppointmentProcessor.java]: done

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/JudicialOfficeAppointmentProcessor.java [hunk 4]
```diff
@@ -39,34 +39,29 @@ public void process(Exchange exchange) {
         List<JudicialOfficeAppointment> filteredJudicialAppointments = validate(judicialOfficeAppointmentJsrValidatorInitializer,
                 judicialOfficeAppointments);
 
-        log.info("Judicial Appointment Records count after Validation:: " + filteredJudicialAppointments.size());
+        List<JudicialUserProfile> invalidJudicialUserProfileRecords = judicialUserProfileJsrValidatorInitializer.getInvalidJsrRecords();
+
+        filterInvalidUserProfileRecords(filteredJudicialAppointments, invalidJudicialUserProfileRecords);
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/JudicialOfficeAppointmentProcessor.java:45]: this should log teh records which are skipped from appointment table in exception table as well

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/util/MappingConstants.java [hunk 5]
```diff
@@ -67,4 +67,7 @@ private MappingConstants() {
 
     public static final String ERROR_MESSAGE = "ErrorMessage";
 
+    public static final String INVALID_JSR_PARENT = "Skipped record as it is invalid in parent";
+
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/util/MappingConstants.java:71]: can u change message to - 
"Record skipped due to jsr violation in the record in the parent load"

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/validator/JsrValidatorInitializer.java [hunk 6]
```diff
@@ -146,6 +147,47 @@ public void setValues(PreparedStatement ps, ConstraintViolation<T> argument) thr
...
+        def.setPropagationBehavior(TransactionDefinition.PROPAGATION_REQUIRED);
+        RouteProperties routeProperties = (RouteProperties) exchange.getIn().getHeader(ROUTE_DETAILS);
+        Map<String, String> globalOptions = camelContext.getGlobalOptions();
+        String schedulerTime = globalOptions.get(SCHEDULER_START_TIME);
+
+        jdbcTemplate.batchUpdate(
+                invalidJsrSql,
+                keys,
+                10,
+                new ParameterizedPreparedStatementSetter<String>() {
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/validator/JsrValidatorInitializer.java:173]: please move batch size property to application.yaml
And batch size can be incrreased.. 10 seems to be too small for batch

Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/juddata/camel/validator/JsrValidatorInitializer.java:176]: no it works efficiently...keeping higher limit sometimes lower performance

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/util/MappingConstants.java (hunk 5 above)
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/juddata/camel/util/MappingConstants.java:71]: done

**Pull request merged**